core-backend/wasmtime/src/env.rs

```diff
@@ -86,12 +115,17 @@ impl<E: Ext + IntoExtInfo> Environment<E> for WasmtimeEnvironment<E> {
         let mut store = Store::<StoreData<E>>::new(&engine, store_data);
 
         // Creates new wasm memory
-        let mut memory = WasmtimeMemory::new(&mut store, MemoryType::new(mem_size.0, None))
-            .map_err(|e| BackendError {
-                reason: "Create env memory failed",
-                description: Some(e.to_string().into()),
-                gas_amount: later_ext.unset().into_gas_amount(),
-            })?;
+        let mut memory = match WasmtimeMemory::new(&mut store, MemoryType::new(mem_size.0, None)) {
```

Comment:
Why `match` style is used across the code instead of previous `?` one?

Reply:
As you can see, we made `LaterExt` consumable in `unset` (which is now `take`). Closure in `map_err` (even `or_else`, which is considered lazy) captures the value and moves it when we try to unset the `LaterExt`. Although it happens inside the err block and immediately returned with `?`, rust can't recognise that the value inside `map_err` closure won't be used later. We have to tell it explicitly with matching. Will be happy if you suggest other neat ways of reaching same aim.

Reply:
https://github.com/gear-tech/gear/pull/836#discussion_r849439289

Reply:
Can we change `self` to `&mut self` in `take` then?

Reply:
It's exactly what was before. It opened ways to unexpected errors and no guarantees of safe usage of the `LaterExt`. Now trying to use type level safety guarantees to free the code of unexpected panics.

Reply:
`unset` returned `E` before, now it is `Option<E>`. I don't get it. So you just panic at caller level (not at `unset` level as before)

Reply:
I understand your point, but if you read the conversation, you will see, that in the next PR (which is stated in code in Todo comments) it's planned to remove `Option` by wiping of the `Clone`. At least for `Sandbox` backend.

Reply:
@ark0f see #853